### PR TITLE
feat/build core erc4626-connector

### PR DESCRIPTION
**Pull request by letteldream**
Build core erc4626 connector.
Available call functions: deposit, mint, withdraw & redeem in erc4626 token contract.

image

### CODE REVIEW

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		IERC4626 tokenContract = IERC4626(token);
+
+		address assetToken = tokenContract.asset();
+		TokenInterface assetTokenContract = TokenInterface(assetToken);
+
+		_amt = _amt == uint256(-1) ? assetTokenContract.balanceOf(token) : _amt;
+
+		approve(assetTokenContract, token, _amt);
+
+		tokenContract.deposit(_amt, msg.sender);
```
Comment (shriyatyagii) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol:42]:
```suggestion
		tokenContract.deposit(_amt, address(this));

```

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		setUint(setId, _amt);
+
+		_eventName = "LogDeposit(address,uint256,uint256,uint256)";
+		_eventParam = abi.encode(token, _amt, getId, setId);
+	}
```
Comment (shriyatyagii) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol:73]: Apply the above deposit function changes to other functions as well.

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		uint256 _amt = getUint(getId, amt);
+		IERC4626 vaultTokenContract = IERC4626(token);
+
+		address _underlyingToken = vaultTokenContract.asset();
+		TokenInterface underlyingTokenContract = TokenInterface(_underlyingToken);
+
+		_amt = _amt == uint256(-1) ? underlyingTokenContract.balanceOf(address(this)) : _amt;
+
+		approve(underlyingTokenContract, token, _amt);
+
```
Comment (KaymasJain) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol:61]: Get balanceOf vault token

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		vaultTokenContract.deposit(_amt, address(this));
+		setUint(setId, _amt);
```
Comment (KaymasJain) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol:43]: Get final balance of vault token.
require(minShares < (final balance - initial balance))

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	 * @notice Mints vault shares by minting exactly amount of underlying assets
+	 * @param token ERC4626 Token address.
+	 * @param amt The amount of the token to mint. (For max: `uint256(-1)`)
+	 * @param getId ID to retrieve amt.
+	 * @param setId ID stores the amount of tokens minted.
+	 */
+
+	function mint(
+		address token,
+		uint256 amt,
```
Comment (KaymasJain) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol]:
```suggestion
		uint256 amt,
		uint256 maxUnderlyingAmt,

```

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+			_underlyingToken
+		);
+
+		_underlyingAmt = _underlyingAmt == uint256(-1)
+			? underlyingTokenContract.balanceOf(address(this))
+			: _underlyingAmt;
+
+		uint256 _minShares = convert18ToDec(
+			_vaultShareDecimal,
+			wmul(minSharesPerToken, _underlyingAmt)
```
Comment (KaymasJain) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol]: Confirming here. `minSharesPerToken` will be in 18 decimals?

File: contracts/mainnet/connectors/basic-ERC4626/main.sol [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+		);
+
+		uint256 _initalVaultBal = vaultTokenContract.balanceOf(address(this));
```
Comment (KaymasJain) [inline contracts/mainnet/connectors/basic-ERC4626/main.sol:52]:
```suggestion
		uint256 _initialVaultBal = vaultTokenContract.balanceOf(address(this));

```

Comment (KaymasJain) [reply contracts/mainnet/connectors/basic-ERC4626/main.sol:52]: Minor typo everywhere

**Review by shriyatyagii — DISMISSED** [review]

**Review by shriyatyagii — APPROVED** [review]

**Pull request merged**